aggregator/utils/utils.py

```diff
@@ -0,0 +1,294 @@
+"""Small General-Purpose Utility Functions."""
+
+import os
+import sys
+import json
+import ssl
+
+import aiohttp
+
+from aiohttp import web
+from aiocache import cached, SimpleMemoryCache
+from aiocache.serializers import JsonSerializer
+
+from ..config import CONFIG
+from .logging import LOG
+
+
+async def http_get_service_urls(registry):
+    """Query an external registry for known service urls of desired type."""
+    LOG.debug('Query external registry for given service type.')
+    service_urls = []
+
+    # Query Registry for services
+    async with aiohttp.ClientSession() as session:
+        try:
+            async with session.get(registry,
+                                   ssl=await request_security()) as response:
+                if response.status == 200:
+                    result = await response.json()
+                    for r in result:
+                        # Parse type `org.ga4gh:service:version`
+                        type_bundle = r.get('type').split(':')
+                        service_type = f'{type_bundle[0]}:{type_bundle[1]}'
+                        # Check if service has a type tag of Beacons
+                        if CONFIG.beacons and service_type == 'org.ga4gh:beacon':
+                            service_urls.append(r['url'])
+                        # Check if service has a type tag of Aggregators
+                        if CONFIG.aggregators and service_type == 'org.ga4gh:beacon-aggregator':
+                            service_urls.append(r['url'])
+        except Exception as e:
+            LOG.debug(f'Query error {e}.')
+            web.HTTPInternalServerError(text=f'An error occurred while attempting to query services: {e}')
+
+    return service_urls
+
+
+# Cache Beacon URLs if they're not already cached
+@cached(ttl=86400, key="beacon_urls", serializer=JsonSerializer())
+async def get_services(url_self):
+    """Return service urls."""
+    LOG.debug('Fetch service urls.')
+
+    # Query Registries for their known Beacon services, fetch only URLs
+    service_urls = set()
+    for registry in CONFIG.registries:
+        services = await http_get_service_urls(registry.get('url', ''))  # Request URLs from Registry
+        service_urls.update(services)  # Add found URLs to set (eliminate duplicates)
+
+    # Pre-process URLS
+    service_urls = [await process_url(url) for url in service_urls]
+    service_urls = await remove_self(url_self, service_urls)

```

Comment:
seems like lines 60+61 might make use of: https://docs.python.org/3/library/itertools.html#itertools.filterfalse